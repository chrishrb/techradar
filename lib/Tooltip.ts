import { Selection, select } from "d3";

export default class Tooltip {
  // TODO
  private tooltip: Selection<HTMLDivElement, unknown, HTMLElement, null>;
  private tooltipSpan: Selection<HTMLSpanElement, unknown, HTMLElement, null>;

  constructor() {
    //create tooltip
    this.tooltip = select("body")
      .append("div")
      .style("font-size", "12px")
      .style("color", "white")
      .style("opacity", 0)
      .style("position", "fixed")
      .style("pointer-events", "none")
      .style("border-radius", "3px")
      .style("background-color", "black")
      .style("padding", "5px 8px");

    //add tooltip's text container
    this.tooltipSpan = this.tooltip
      .append("span")
      .style("white-space", "nowrap");

    //add tooltip's box arrow
    this.tooltip
      .append("div")
      .text("â–¼")
      .style("color", "black")
      .style("position", "absolute")
      .style("margin-top", "-3px")
      .style("top", "100%")
      .style("left", 0)
      .style("width", "100%")
      .style("text-align", "center");
  }

  show = (text: string, anchorX: number, anchorY: number) => {
    this.tooltipSpan.text(text);
    const node = this.tooltip.node();
    this.tooltip
      .style("opacity", 1)
      .style("left", anchorX - (node ? node.offsetWidth : 0) / 2 + "px")
      .style("top", anchorY - (node ? node.offsetHeight : 0) - 10 + "px");
  };

  hide = () => {
    this.tooltip.style("opacity", 0);
    this.tooltipSpan.text("");
  };

  destroy = () => {
    this.tooltip.remove();
  };
}
