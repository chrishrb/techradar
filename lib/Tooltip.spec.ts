import { describe, expect, beforeEach, it, afterEach } from 'vitest'
import Tooltip from "./Tooltip";

describe("Tooltip", () => {
  let tooltip: Tooltip;

  const text = "Hello!";
  const anchorX = 500;
  const anchorY = 1000;

  beforeEach(() => {
    tooltip = new Tooltip();
  });

  afterEach(() => {
    document.body.innerHTML = "";
  });

  it("should inject a hidden div into the body when created", () => {
    const tooltipEl = document.querySelector("body > div") as HTMLDivElement;

    expect(tooltipEl.style.getPropertyValue("opacity")).toEqual("0");

    expect(tooltipEl).toMatchSnapshot();
  });

  it("should make tooltip visible with specified text, on top of a certain anchor point, after calling .show()", () => {
    tooltip.show(text, anchorX, anchorY);

    const tooltipEl = document.querySelector("body > div") as HTMLDivElement;

    expect(tooltipEl.style.getPropertyValue("opacity")).toEqual("1");

    //horizontally centered on the anchorX
    const expectedLeft = anchorX - tooltipEl.offsetWidth / 2 + "px";
    expect(tooltipEl.style.getPropertyValue("left")).toEqual(expectedLeft);

    //vertically above the anchorY (with some margin)
    const expectedTop = anchorY - tooltipEl.offsetHeight - 10 + "px";
    expect(tooltipEl.style.getPropertyValue("top")).toEqual(expectedTop);

    const tooltipSpanEl = tooltipEl.querySelector("span");
    expect(tooltipSpanEl?.textContent).toBe(text);
  });

  it("should hide the tooltip, after calling .show() and then .hide()", () => {
    tooltip.show(text, anchorX, anchorY);
    tooltip.hide();

    const tooltipEl = document.querySelector("body > div") as HTMLDivElement;
    expect(tooltipEl.style.getPropertyValue("opacity")).toEqual("0");
  });

  it("should remove the tooltip, after calling .destroy()", () => {
    tooltip.destroy();

    const tooltipEl = document.querySelector("body > div");
    expect(tooltipEl).toBeNull();
  });
});
