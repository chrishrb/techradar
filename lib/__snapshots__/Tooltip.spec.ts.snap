// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip > should inject a hidden div into the body when created 1`] = `
<div
  style="font-size: 12px; color: white; opacity: 0; position: fixed; pointer-events: none; border-radius: 3px; background-color: black; padding: 5px 8px;"
>
  <span
    style="white-space: nowrap;"
  />
  <div
    style="color: black; position: absolute; margin-top: -3px; top: 100%; left: 0px; width: 100%; text-align: center;"
  >
    ▼
  </div>
</div>
`;

exports[`Tooltip should inject a hidden div into the body when created 1`] = `
<div
  style="font-size: 12px; color: white; opacity: 0; position: fixed; pointer-events: none; border-radius: 3px; background-color: black; padding: 5px 8px;"
>
  <span
    style="white-space: nowrap;"
  />
  <div
    style="color: black; position: absolute; margin-top: -3px; top: 100%; left: 0px; width: 100%; text-align: center;"
  >
    ▼
  </div>
</div>
`;
